File: 2023/day_1/main.py
def part_1() -> None:
    with open("input.txt") as f:
        lines: list[str] = f.read().splitlines()

    sum: int = 0
    for line in lines:
        first: bool = False
        first_digit: str = ""
        last_digit: str = ""

        for char in line:
            if char.isdigit() and not first:
                first_digit = char
                first = True
            elif char.isdigit():
                last_digit = char

        if last_digit == "":
            last_digit = first_digit

        combined: str = first_digit + last_digit

        sum += int(combined)

    print(sum)


def part_2() -> None:
    with open("input.txt") as f:
        lines: list[str] = f.read().splitlines()

    map: dict[str, str] = {
        "one": "one1one",
        "two": "two2two",
        "three": "three3three",
        "four": "four4four",
        "five": "five5five",
        "six": "six6six",
        "seven": "seven7seven",
        "eight": "eight8eight",
        "nine": "nine9nine",
    }

    sum: int = 0
    for line in lines:
        print(line)
        for key, value in map.items():
            line = line.replace(key, value)

        line = "".join(filter(str.isdigit, line))
        sum += int(line[0]+line[-1])

    print(sum)


if __name__ == "__main__":
    part_1()
    part_2()

File: 2023/day_1/main.swift
import Foundation

func readFile(atPath path: String) -> String? {
    do {
        let contents = try String(contentsOfFile: path)
        return contents
    } catch {
        print("Unable to read file: \(error)")
        return nil
    }
}

func isDigit(_ char: Character) -> Bool {
    return char >= "0" && char <= "9"
}

func part1() {
    let contents = readFile(atPath: "input.txt")!
        let lines = contents.split(separator: "\n")

        var sum = 0
        for line in lines {
            var firstDigit = ""
                var lastDigit = ""
                var foundFirstDigit = false

                for char in line {
                    let isCharDigit = isDigit(char)
                        if isCharDigit && !foundFirstDigit {
                            firstDigit = String(char)
                                foundFirstDigit = true
                        } else if isCharDigit {
                            lastDigit = String(char)
                        }
                }

            if lastDigit == "" {
                lastDigit = firstDigit
            }

            sum += Int(firstDigit + lastDigit)!
        }

    print(sum)
}

func part2() {
    let contents = readFile(atPath: "input.txt")!
    let lines = contents.split(separator: "\n")
    
    let map = [
        "one": "one1one",
        "two": "two2two",
        "three": "three3three",
        "four": "four4four",
        "five": "five5five",
        "six": "six6six",
        "seven": "seven7seven",
        "eight": "eight8eight",
        "nine": "nine9nine",
    ]

    var sum = 0
    for line in lines {
        var newLine = String(line)
        for (key, value) in map {
            newLine = newLine.replacingOccurrences(of: key, with: value)
        }
        newLine = newLine.replacingOccurrences(of: "[^1-9]", with: "", options: .regularExpression)
        sum += Int(newLine.prefix(1) + newLine.suffix(1))!
    }
    print(sum)
}

part1()
part2()
